### PR TITLE
MTDSA-26817: Cumulative new end point unit

### PR DESCRIPTION
**Pull request by iyadi**

### CODE REVIEW

File: app/v4/createAmendCumulativeSubmission/CreateAmendCumulativeSubmissionConnector.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        val downstreamUri =
+          uriFactory(nino, businessId, taxYear)
+        put(body, downstreamUri)
+    }
+
+  }
+
+  private def uriFactory(nino: Nino, businessId: BusinessId, taxYear: TaxYear) = {
+    if (taxYear.useTaxYearSpecificApi) {
+      TaxYearSpecificIfsUri[Unit](s"income-tax/${taxYear.asTysDownstream}/$nino/self-employments/$businessId/cumulative-summaries")
```
Comment (FergusHMRC) [inline app/v4/createAmendCumulativeSubmission/CreateAmendCumulativeSubmissionConnector.scala:53]: URLs here are incorrect, should be s"income-tax/${taxYear.asTysDownstream}/self-employments/periodic/$nino/$businessId" to match the URL in the technical requirements

File: app/api/models/errors/SeBusinessCommonErrors.scala [hunk 2]
```diff
@@ -90,3 +90,38 @@ object RuleWrongTpaAmountSubmittedError
       code = "RULE_WRONG_TPA_AMOUNT_SUBMITTED",
       message = "Transition profit acceleration value cannot be submitted without a transition profit value",
       BAD_REQUEST)
+
```
Comment (FergusHMRC) [inline app/api/models/errors/SeBusinessCommonErrors.scala:93]: Error messages dont match the error messages in the business requirements, check them and correct

**Review by FergusHMRC — CHANGES_REQUESTED** [review]

File: app/api/models/errors/SeBusinessCommonErrors.scala (hunk 2 above)
Comment (iyadi) [reply app/api/models/errors/SeBusinessCommonErrors.scala:93]: this is an existing code and not part of the PR

File: app/api/models/errors/SeBusinessCommonErrors.scala [hunk 3]
```diff
@@ -90,3 +90,38 @@ object RuleWrongTpaAmountSubmittedError
...
+object RuleEndDateNotAlignedWithReportingTypeError
+    extends MtdError(code = "RULE_END_DATE_NOT_ALIGNED_WITH_REPORTING_TYPE", message = "End date is not aligned with the reporting type", BAD_REQUEST)
+
+object RuleMissingSubmissionDatesError extends MtdError(code = "RULE_MISSING_SUBMISSION_DATES", message = "Submission dates are missing", BAD_REQUEST)
+
+object RuleStartAndEndDateNotAllowedError
+    extends MtdError(code = "RULE_START_AND_END_DATE_NOT_ALLOWED", message = "Start and end dates are not allowed", BAD_REQUEST)
+
+object RuleOutsideAmendmentWindowError
+    extends MtdError(code = "RULE_OUTSIDE_AMENDMENT_WINDOW", message = "Submission is outside the amendment window", BAD_REQUEST)
```
Comment (FergusHMRC) [inline app/api/models/errors/SeBusinessCommonErrors.scala:127]: error messages dont match tech specs

File: app/v4/createAmendCumulativeSubmission/CreateAmendCumulativeSubmissionConnector.scala [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        put(body, downstreamUri)
+    }
+
+  }
+
+  private def uriFactory(nino: Nino, businessId: BusinessId, taxYear: TaxYear) = {
+    if (taxYear.useTaxYearSpecificApi) {
+      TaxYearSpecificIfsUri[Unit](s"income-tax/${taxYear.asTysDownstream}/self-employments/periodic/$nino/$businessId")
+    } else {
+      IfsUri[Unit](s"income-tax/nino/$nino/self-employments/$businessId/cumulative-summaries/${taxYear.asDownstream}")
```
Comment (FergusHMRC) [inline app/v4/createAmendCumulativeSubmission/CreateAmendCumulativeSubmissionConnector.scala:55]: IfsUri and the check can be removed, all pre TYS tax years will return an tax year not supported error

File: app/v4/createAmendCumulativeSubmission/CreateAmendCumulativeSubmissionService.scala [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      "EARLY_DATA_SUBMISSION_NOT_ACCEPTED" -> RuleEarlyDataSubmissionNotAcceptedError,
+      "INVALID_SUBMISSION_END_DATE"        -> RuleAdvanceSubmissionRequiresPeriodEndDateError,
+      "SUBMISSION_END_DATE_VALUE"          -> RuleSubmissionEndDateCannotMoveBackwardsError,
+      "INVALID_START_DATE"                 -> RuleStartDateNotAlignedWithReportingTypeError,
+      "START_DATE_NOT_ALIGNED"             -> RuleStartDateNotAlignedToCommencementDateError,
+      "END_DATE_NOT_ALIGNED"               -> RuleEndDateNotAlignedWithReportingTypeError,
+      "MISSING_SUBMISSION_DATES"           -> RuleMissingSubmissionDatesError,
+      "START_END_DATE_NOT_ACCEPTED"        -> RuleStartAndEndDateNotAllowedError,
+      "OUTSIDE_AMENDMENT_WINDOW"           -> RuleOutsideAmendmentWindowError
+    )
```
Comment (FergusHMRC) [inline app/v4/createAmendCumulativeSubmission/CreateAmendCumulativeSubmissionService.scala:66]: Missing BOTH_EXPENSES_SUPPLIED here

File: app/v4/createAmendCumulativeSubmission/CreateAmendCumulativeSubmissionConnector.scala [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import shared.config.SharedAppConfig
+import play.api.http.Status.OK
+import uk.gov.hmrc.http.{HeaderCarrier, HttpClient}
+import v4.createAmendCumulativeSubmission.model.request.{CreateAmendCumulativeSubmissionRequestData, Def1_CreateAmendCumulativeSubmissionRequestData}
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class CreateAmendCumulativeSubmissionConnector @Inject() (val http: HttpClient, val appConfig: SharedAppConfig) extends BaseDownstreamConnector {
```
Comment (GazalNebHMRC) [inline app/v4/createAmendCumulativeSubmission/CreateAmendCumulativeSubmissionConnector.scala:32]: The endpoint should be named CreateAmendCumulativePeriodSummaryConnector

File: app/api/models/errors/SeBusinessCommonErrors.scala [hunk 7]
```diff
@@ -90,3 +90,48 @@ object RuleWrongTpaAmountSubmittedError
...
+    extends MtdError(
+      code = "RULE_START_DATE_NOT_ALIGNED_TO_COMMENCEMENT_DATE",
+      message = "Start date does not align to the commencement date.",
+      BAD_REQUEST)
+
+object RuleEndDateNotAlignedWithReportingTypeError
+    extends MtdError(code = "RULE_END_DATE_NOT_ALIGNED_WITH_REPORTING_TYPE", message = "End date does not align to the reporting type.", BAD_REQUEST)
+
+object RuleMissingPeriodSummaryDatesError
+    extends MtdError(code = "RULE_MISSING_SUBMISSION_DATES", message = "PeriodSummary start/end date not provided", BAD_REQUEST)
```
Comment (FergusHMRC) [inline app/api/models/errors/SeBusinessCommonErrors.scala:128]: Change "PeriodSummary start/end date not provided" to "Submission start/end date not provided."

File: test/v4/createAmendCumulativePeriodSummary/CreateAmendCumulativePeriodSummaryConnectorSpec.scala [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    )
+
+    def taxYear: TaxYear
+
+  }
+
+  "AmendCumulativePeriodSummaryConnector for a Tax Year Specific tax year" must {
+    "return a 200 status for a success scenario" in
+      new TysIfsTest with Test {
+        def taxYear: TaxYear = TaxYear.fromMtd("2023-24")
```
Comment (FergusHMRC) [inline test/v4/createAmendCumulativePeriodSummary/CreateAmendCumulativePeriodSummaryConnectorSpec.scala:65]: shouldnt this fail as its before the 2025-26 minimum date?

Comment (iyadi) [reply test/v4/createAmendCumulativePeriodSummary/CreateAmendCumulativePeriodSummaryConnectorSpec.scala:65]: done

File: app/api/models/errors/SeBusinessCommonErrors.scala (hunk 7 above)
Comment (iyadi) [reply app/api/models/errors/SeBusinessCommonErrors.scala:128]: done

File: test/v4/createAmendCumulativePeriodSummary/def1/Def1_CreatePeriodSummaryValidatorSpec.scala [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+  private val validNino       = "AA123456A"
+  private val validBusinessId = "XAIS12345678901"
+  private val taxYear         = TaxYear.fromMtd("2025-26")
+
+  private val validPeriodDates = Json.parse("""
+      |{
+      |  "periodStartDate": "2025-08-24",
+      |  "periodEndDate": "2026-08-24"
+      |}
+      |""".stripMargin)
```
[the remark this answers is not in the document]
Comment (iyadi) [reply test/v4/createAmendCumulativePeriodSummary/def1/Def1_CreatePeriodSummaryValidatorSpec.scala:47]: done and fixed

File: it/v4/createAmendCumulativePeriodSummary/def1/Def1_CreateAmendCumulativePeriodSummaryControllerISpec.scala [hunk 10]
```diff
@@ -0,0 +1,441 @@
...
+        (BAD_REQUEST, "INVALID_NINO", BAD_REQUEST, NinoFormatError),
+        (BAD_REQUEST, "INVALID_INCOME_SOURCE", BAD_REQUEST, BusinessIdFormatError),
+        (BAD_REQUEST, "INVALID_PAYLOAD", INTERNAL_SERVER_ERROR, InternalError),
+        (NOT_FOUND, "NOT_FOUND_INCOME_SOURCE", NOT_FOUND, NotFoundError),
+        (INTERNAL_SERVER_ERROR, "SERVER_ERROR", INTERNAL_SERVER_ERROR, InternalError),
+        (SERVICE_UNAVAILABLE, "SERVICE_UNAVAILABLE", INTERNAL_SERVER_ERROR, InternalError),
+        (BAD_REQUEST, "INVALID_CORRELATIONID", INTERNAL_SERVER_ERROR, InternalError),
+        (BAD_REQUEST, "INVALID_INCOME_SOURCE_ID", BAD_REQUEST, BusinessIdFormatError),
+        (UNPROCESSABLE_ENTITY, "TAX_YEAR_NOT_SUPPORTED", BAD_REQUEST, RuleTaxYearNotSupportedError),
+        (NOT_FOUND, "INCOME_SOURCE_NOT_FOUND", NOT_FOUND, NotFoundError)
```
Comment (FergusHMRC) [inline it/v4/createAmendCumulativePeriodSummary/def1/Def1_CreateAmendCumulativePeriodSummaryControllerISpec.scala:346]: missing downstream errors from CreateAmendCumulativePeriodSummaryService

File: app/v4/createAmendCumulativePeriodSummary/CreateAmendCumulativePeriodSummaryService.scala [hunk 11]
```diff
@@ -62,15 +53,16 @@ class CreateAmendCumulativePeriodSummaryService @Inject() (connector: CreateAmen
       "END_DATE_NOT_ALIGNED"               -> RuleEndDateNotAlignedWithReportingTypeError,
       "MISSING_SUBMISSION_DATES"           -> RuleMissingPeriodSummaryDatesError,
       "START_END_DATE_NOT_ACCEPTED"        -> RuleStartAndEndDateNotAllowedError,
-      "OUTSIDE_AMENDMENT_WINDOW"           -> RuleOutsideAmendmentWindowError
+      "OUTSIDE_AMENDMENT_WINDOW"           -> RuleOutsideAmendmentWindowError,
+      "SERVER_ERROR"                       -> InternalError,
+      "SERVICE_UNAVAILABLE"                -> InternalError
     )
 
     val extraTysErrors = Map(
```
Comment (FergusHMRC) [inline app/v4/createAmendCumulativePeriodSummary/CreateAmendCumulativePeriodSummaryService.scala:61]: remove these 3 lines and put the errors below in a suitable place before the API specific errors

File: it/v4/createAmendCumulativePeriodSummary/def1/Def1_CreateAmendCumulativePeriodSummaryControllerISpec.scala [hunk 12]
```diff
@@ -17,6 +17,18 @@
...
+  RuleBothExpensesSuppliedError,
+  RuleEarlyDataPeriodSummaryNotAcceptedError,
+  RuleEndDateNotAlignedWithReportingTypeError,
+  RuleMissingPeriodSummaryDatesError,
+  RuleOutsideAmendmentWindowError,
+  RulePeriodSummaryEndDateCannotMoveBackwardsError,
+  RuleStartAndEndDateNotAllowedError,
+  RuleStartDateNotAlignedToCommencementDateError,
+  RuleStartDateNotAlignedWithReportingTypeError
+}
```
Comment (FergusHMRC) [inline it/v4/createAmendCumulativePeriodSummary/def1/Def1_CreateAmendCumulativePeriodSummaryControllerISpec.scala:31]: dont leave commented out code inside the code, in this case errors._ is better as you are importing pretty much all of them anyway.

**Comment by FergusHMRC** [comment]
test this please

**Pull request merged**